# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [1]:
# Import your libraries:
import sklearn
from sklearn import datasets
import pandas as pd
import numpy as np

# Challenge 1 - Explore the Scikit-Learn Datasets

Before starting to work on our own datasets, let's first explore the datasets that are included in this Python library. These datasets have been cleaned and formatted for use in ML algorithms.

First, we will load the diabetes dataset. Do this in the cell below by importing the datasets and then loading the dataset  to the `diabetes` variable using the `load_diabetes()` function ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_diabetes.html)).

In [2]:
# Your code here:
diabetes = datasets.load_diabetes()

Let's explore this variable by looking at the different attributes (keys) of `diabetes`. Note that the `load_diabetes` function does not return dataframes. It returns you a Python dictionary.

In [3]:
# Your code here:
diabetes.keys()

dict_keys(['data', 'target', 'DESCR', 'feature_names', 'data_filename', 'target_filename'])

#### The next step is to read the description of the dataset. 

Print the description in the cell below using the `DESCR` attribute of the `diabetes` variable. Read the data description carefully to fully understand what each column represents.

*Hint: If your output is ill-formatted by displaying linebreaks as `\n`, it means you are not using the `print` function.*

In [4]:
# Your code here:
print(diabetes.DESCR)


.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - Age
      - Sex
      - Body mass index
      - Average blood pressure
      - S1
      - S2
      - S3
      - S4
      - S5
      - S6

Note: Each of these 10 feature variables have been mean centered and scaled by the standard deviation times `n_samples` (i.e. the sum of squares of each column totals 1).

Source URL:
https://www4.stat.ncsu.edu/~boos/var.select/diabetes.html

For more information see:
Bra

#### Based on the data description, answer the following questions:

1. How many attributes are there in the data? What do they mean?

1. What is the relation between `diabetes['data']` and `diabetes['target']`?

1. How many records are there in the data?

# Enter your answer here:
1.hay 10 atributos(columnas)
2.Target es lo que avanzo la enfermedad desde que se inicia hasta pasado un año
3.Numero de instancias : 442


#### Now explore what are contained in the *data* portion as well as the *target* portion of `diabetes`. 

Scikit-learn typically takes in 2D numpy arrays as input (though pandas dataframes are also accepted). Inspect the shape of `data` and `target`. Confirm they are consistent with the data description.

In [5]:
# Your code here:
print(diabetes['data'].shape, diabetes['target'].shape)
diabetes_df = pd.DataFrame(diabetes['data'])
diabetes_df.head()

(442, 10) (442,)


0         1         2         3         4         5         6  \
0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1 -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2  0.085299  0.050680  0.044451 -0.005671 -0.045599 -0.034194 -0.032356   
3 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   

          7         8         9  
0 -0.002592  0.019908 -0.017646  
1 -0.039493 -0.068330 -0.092204  
2 -0.002592  0.002864 -0.025930  
3  0.034309  0.022692 -0.009362  
4 -0.002592 -0.031991 -0.046641

In [6]:
target_df = pd.DataFrame(diabetes['target'])
target_df.head()

0
0  151.0
1   75.0
2  141.0
3  206.0
4  135.0

# Challenge 2 - Perform Supervised Learning on the Dataset

The data have already been split to predictor (*data*) and response (*target*) variables. Given this information, we'll apply what we have previously learned about linear regression and apply the algorithm to the diabetes dataset.

#### Let's briefly revisit the linear regression formula:

```
y = β0 + β1X1 + β2X2 + ... + βnXn + ϵ
```

...where:

- X1-Xn: data 
- β0: intercept 
- β1-βn: coefficients 
- ϵ: error (cannot explained by model)
- y: target

Also take a look at the `sklearn.linear_model.LinearRegression` [documentation](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html).

#### In the cell below, import the `linear_model` class from `sklearn`. 

In [7]:
# Your code here:
from sklearn import linear_model
#from sklearn.linear_model import LinearRegression

#### Create a new instance of the linear regression model and assign the new instance to the variable `diabetes_model`.

In [8]:
# Your code here:
diabetes_model= linear_model.LinearRegression()

#### Next, let's split the training and test data.

Define `diabetes_data_train`, `diabetes_target_train`, `diabetes_data_test`, and `diabetes_target_test`. Use the last 20 records for the test data and the rest for the training data.

In [9]:
# Your code here:
diabetes_data_train = diabetes_df[:-20]
diabetes_target_train = target_df[:-20]
diabetes_data_test = diabetes_df[-20:]
diabetes_target_test = target_df[-20:]

Fit the training data and target to `diabetes_model`. Print the *intercept* and *coefficients* of the model.

In [10]:
# Your code here:
a=diabetes_model.fit(diabetes_data_train,diabetes_target_train)
print(a.coef_)
print(a.intercept_)

[[ 3.03499549e-01 -2.37639315e+02  5.10530605e+02  3.27736980e+02
  -8.14131709e+02  4.92814588e+02  1.02848452e+02  1.84606489e+02
   7.43519617e+02  7.60951722e+01]]
[152.76430692]


#### Inspecting the results

From the outputs you should have seen:

- The intercept is a float number.
- The coefficients are an array containing 10 float numbers.

This is the linear regression model fitted to your training dataset.

#### Using your fitted linear regression model, predict the *y* of `diabetes_data_test`.

In [11]:
# Your code here:
y_pred = a.predict(diabetes_data_test)
print(y_pred)

[[197.61846908]
 [155.43979328]
 [172.88665147]
 [111.53537279]
 [164.80054784]
 [131.06954875]
 [259.12237761]
 [100.47935157]
 [117.0601052 ]
 [124.30503555]
 [218.36632793]
 [ 61.19831284]
 [132.25046751]
 [120.3332925 ]
 [ 52.54458691]
 [194.03798088]
 [102.57139702]
 [123.56604987]
 [211.0346317 ]
 [ 52.60335674]]


#### Print your `diabetes_target_test` and compare with the prediction. 

In [12]:
# Your code here:
print(diabetes_target_test)

         0
422  233.0
423   91.0
424  111.0
425  152.0
426  120.0
427   67.0
428  310.0
429   94.0
430  183.0
431   66.0
432  173.0
433   72.0
434   49.0
435   64.0
436   48.0
437  178.0
438  104.0
439  132.0
440  220.0
441   57.0


#### Is `diabetes_target_test` exactly the same as the model prediction? Explain.

In [13]:
# Your explanation here:
diabetes_target_test['pred']=y_pred
diabetes_target_test

/usr/lib/python3/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


0        pred
422  233.0  197.618469
423   91.0  155.439793
424  111.0  172.886651
425  152.0  111.535373
426  120.0  164.800548
427   67.0  131.069549
428  310.0  259.122378
429   94.0  100.479352
430  183.0  117.060105
431   66.0  124.305036
432  173.0  218.366328
433   72.0   61.198313
434   49.0  132.250468
435   64.0  120.333292
436   48.0   52.544587
437  178.0  194.037981
438  104.0  102.571397
439  132.0  123.566050
440  220.0  211.034632
441   57.0   52.603357

No son iguales, tiene errores

# Bonus Challenge 1 - Hypothesis Testing with `statsmodels`

After generating the linear regression model from the dataset, you probably wonder: then what? What is the statistical way to know if my model is reliable or not?

Good question. We'll discuss that using Scikit-Learn in Challenge 5. But for now, let's use a fool-proof way by using the ([Linear Regression class of StatsModels](https://www.statsmodels.org/dev/regression.html)) which can also conduct linear regression analysis plus much more such as calcuating the F-score of the linear model as well as the standard errors and t-scores for each coefficient. The F-score and t-scores will tell you whether you can trust your linear model.

To understand the statistical meaning of conducting hypothesis testing (e.g. F-test, t-test) for slopes, read [this webpage](https://onlinecourses.science.psu.edu/stat501/node/297/) at your leisure time. We'll give you a brief overview next.

* The F-test of your linear model is to verify whether at least one of your coefficients is significantly different from zero. Translating that into the *null hypothesis* and *alternative hypothesis*, that is:

    ```
    H0 : β1 = β2 = ... = β10 = 0
    HA : At least one βj ≠ 0 (for j = 1, 2, ..., 10)
    ```

* The t-tests on each coefficient is to check whether the confidence interval for the variable contains zero. If the confidence interval contains zero, it means the null hypothesis for that variable is not rejected. In other words, this particular vaiable is not contributing to your linear model and you can remove it from your formula.

Read the documentations of [StatsModels Linear Regression](https://www.statsmodels.org/dev/regression.html) as well as its [`OLS` class](https://www.statsmodels.org/dev/generated/statsmodels.regression.linear_model.OLS.html) which stands for *ordinary least squares*.

#### In the next cell, analyze `diabetes_data_train` and `diabetes_target_train` with the linear regression model of `statsmodels`. Print the fit summary.

Your output should look like:

![statsmodels regression](../statsmodels.png)

In [14]:
# Your code here:
import statsmodels.api as sm

diabetes_data_train = sm.add_constant(diabetes_data_train, prepend=False)
mod = sm.OLS(diabetes_target_train, diabetes_data_train)
res=mod.fit()
print(res.summary())

                            OLS Regression Results                            
Dep. Variable:                      0   R-squared:                       0.512
Model:                            OLS   Adj. R-squared:                  0.500
Method:                 Least Squares   F-statistic:                     43.16
Date:                Thu, 27 Feb 2020   Prob (F-statistic):           4.64e-58
Time:                        20:25:29   Log-Likelihood:                -2281.1
No. Observations:                 422   AIC:                             4584.
Df Residuals:                     411   BIC:                             4629.
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
0              0.3035     61.286      0.005      0.9

#### Interpreting hypothesis testing results

Answer the following questions in the cell below:

1. What is the F-score of your linear model and is the null hypothesis rejected?

1. Does any of the t-tests of the coefficients produce a confidence interval containing zero? What are they?

1. How will you modify your linear reguression model according to the test results above?

# Your answers here:
1.Prob (F-statistic):4.64e-58...como es mucho menor que 0.05 podemos rachazar la hipotesis nula, por tanto es posible que sea verdadera

2.mirando en las columnas [0.025 y 0.975]contienen 0 :0,4,5,6,7,9


3.se podrian eliminar los que P>|t|== 0.000 ... 1,2,3 y 8

# Challenge 3 - Peform Supervised Learning on a Pandas Dataframe

Now that we have dealt with data that has been formatted for scikit-learn, let's look at data that we will need to format ourselves.

In the next cell, load the `auto-mpg.csv` file included in this folder and assign it to a variable called `auto`.

In [15]:
# Your code here:
auto = pd.read_csv('../auto-mpg.csv')

Look at the first 5 rows using the `head()` function:

In [16]:
# Your code here:
auto.head()

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   
3  16.0          8         304.0        150.0    3433          12.0   
4  17.0          8         302.0        140.0    3449          10.5   

   model_year                       car_name  
0          70  \t"chevrolet chevelle malibu"  
1          70          \t"buick skylark 320"  
2          70         \t"plymouth satellite"  
3          70              \t"amc rebel sst"  
4          70                \t"ford torino"

Evaluate the data to ensure that all numeric columns are correctly detected as such by pandas. If a column is misclassified as object, coerce it to numeric.

In [17]:
# Your code here:
auto.dtypes

mpg             float64
cylinders         int64
displacement    float64
horse_power     float64
weight            int64
acceleration    float64
model_year        int64
car_name         object
dtype: object

What is the newest model year and the oldest model year?

In [18]:
# Your code here:
print(max(auto['model_year']))
print(min(auto['model_year']))

82
70


Check the dataset for missing values and remove all rows containing at least one missing value.

In [19]:
auto.isna().sum()

mpg             0
cylinders       0
displacement    0
horse_power     6
weight          0
acceleration    0
model_year      0
car_name        0
dtype: int64

In [20]:
# Your code here:

auto.dropna(axis=0, inplace=True)
auto.head()

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   
3  16.0          8         304.0        150.0    3433          12.0   
4  17.0          8         302.0        140.0    3449          10.5   

   model_year                       car_name  
0          70  \t"chevrolet chevelle malibu"  
1          70          \t"buick skylark 320"  
2          70         \t"plymouth satellite"  
3          70              \t"amc rebel sst"  
4          70                \t"ford torino"

Find the frequency table for the `cylinders` column using the `value_counts()` function. How many possible values of cylinders are there?

In [21]:
# Your code here:
auto['cylinders'].value_counts()

4    199
8    103
6     83
3      4
5      3
Name: cylinders, dtype: int64

We would like to generate a linear regression model that will predict mpg. To do this, first drop the `car_name` column since it does not contain any quantitative data. Next separate the dataframe to predictor and response variables. Separate those into test and training data with 80% of the data in the training set and the remainder in the test set. 

Assign the predictor and response training data to `X_train` and `y_train` respectively. Similarly, assign the predictor and response test data to `X_test` and `y_test`.

*Hint: To separate data for training and test, use the `train_test_split` method we used in previous labs.*

In [22]:
# Your code here:
auto_n = auto.drop(['car_name'], axis=1)
auto_n.head()

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   
3  16.0          8         304.0        150.0    3433          12.0   
4  17.0          8         302.0        140.0    3449          10.5   

   model_year  
0          70  
1          70  
2          70  
3          70  
4          70

In [23]:
from sklearn.model_selection import train_test_split
X = auto[['cylinders','displacement','horse_power','weight','acceleration','model_year']]
y = auto[['mpg']]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

Now we will processed and peform linear regression on this data to predict the mpg for each vehicle. 

#### In the next cell, create an instance of the linear regression model and call it `auto_model`. Fit `auto_model` with your training data.

In [24]:
auto_model= linear_model.LinearRegression()
coche = auto_model.fit(X_train,y_train)
print(coche.coef_)
print(coche.intercept_)

[[-0.35791833  0.00865796 -0.00223009 -0.00667633  0.11978438  0.78924994]]
[-17.96388812]


# Challenge 4 - Evaluate the Model

In addition to evaluating your model with F-test and t-test, you can also use the *Coefficient of Determination* (a.k.a. *r squared score*). This method does not simply tell *yes* or *no* about the model fit but instead indicates how much variation can be explained by the model. Based on the r squared score, you can decide whether to improve your model in order to obtain a better fit.

You can learn about the r squared score [here](). Its formula is:

![R Squared](../r-squared.png)

...where:

* yi is an actual data point.
* ŷi is the corresponding data point on the estimated regression line.

By adding the squares of the difference between all yi-ŷi pairs, we have a measure called SSE (*error sum of squares*) which is an application of the r squared score to indicate the extent to which the estimated regression model is different from the actual data. And we attribute that difference to the random error that is unavoidable in the real world. Obviously, we want the SSE value to be as small as possible.

#### In the next cell, compute the predicted *y* based on `X_train` and call it `y_pred`. Then calcualte the r squared score between `y_pred` and `y_train` which indicates how well the estimated regression model fits the training data.

*Hint: r squared score can be calculated using `sklearn.metrics.r2_score` ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html)).*

In [25]:
# Your code here:
y_pred = coche.predict(X_train)

In [26]:
from sklearn.metrics import r2_score
r2_score(y_train, y_pred)

0.8064035638207429

#### Our next step is to evaluate the model using the test data. 

We would like to ensure that our model is not overfitting the data. This means that our model was made to fit too closely to the training data by being overly complex. If a model is overfitted, it is not generalizable to data outside the training data. In that case, we need to reduce the complexity of the model by removing certain features (variables).

In the cell below, use the model to generate the predicted values for the test data and assign them to `y_test_pred`. Compute the r squared score of the predicted `y_test_pred` and the oberserved `y_test` data.

In [27]:
# Your code here:
y_test_pred=coche.predict(X_test)
r2_score(y_test, y_test_pred)


0.8185553501175045

#### Explaining the results

The r squared scores of the training data and the test data are pretty close (0.8146 vs 0.7818). This means our model is not overfitted. However, there is still room to improve the model fit. Move on to the next challenge.

🤬 A mi no me da eso!!! que raaaarooooo... me da 0.8100 vs 0.8023 🤯

# Challenge 5 - Improve the Model Fit

While the most common way to improve the fit of a model is by using [regularization](https://datanice.github.io/machine-learning-101-what-is-regularization-interactive.html), there are other simpler ways to improve model fit. The first is to create a simpler model. The second is to increase the train sample size.

Let us start with the easier option and increase our train sample size to 90% of the data. Create a new test train split and name the new predictors and response variables `X_train09`, `X_test09`, `y_train09`, `y_test09`.

In [28]:
# Your code here:
X_train09, X_test09, y_train09, y_test09 = train_test_split(X, y, test_size=0.1)

Initialize a new linear regression model. Name this model `auto_model09`. Fit the model to the new sample (training) data.

In [29]:
# Your code here:
auto_model09 = linear_model.LinearRegression()
coche09 = auto_model.fit(X_train09,y_train09)
print(coche09.coef_)
print(coche09.intercept_)

[[-0.38602375  0.00539545  0.00711993 -0.0066867   0.13393699  0.76336766]]
[-16.39187342]


Compute the predicted values and r squared score for our new model and new sample data.

In [30]:
# Your code here:
y_pred09 = coche09.predict(X_train09)
r2_score(y_train09, y_pred09)

0.8067485325032864

Compute the r squared score for the smaller test set. Is there an improvement in the test r squared?

In [31]:
# Your code here:
y_test_pred09=coche.predict(X_test09)
r2_score(y_test09, y_test_pred09)

0.8257693754934683

👀En mi caso empeora... 0.8023 vs 0.7710👀

# Bonus Challenge 2 - Backward Elimination 

The main way to produce a simpler linear regression model is to reduce the number of variables used in the model. In scikit-learn, we can do this by using recursive feature elimination. You can read more about RFE [here](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFE.html).

In the next cell, we will import RFE

In [32]:
from sklearn.feature_selection import RFE

Follow the documentation and initialize an RFE model using the `auto_model` linear regression model. Set `n_features_to_select=3`

In [33]:
# Your code here:
selector = RFE(auto_model, n_features_to_select=3)

Fit the model and print the ranking

In [34]:
# Your code here:
selector.fit(X_train, y_train)
selector.ranking_

/home/jose/.local/lib/python3.6/site-packages/sklearn/utils/validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


array([1, 2, 4, 3, 1, 1])

Feature importance is ranked from most important (1) to least important (4). Generate a model with the three most important features. The features correspond to variable names. For example, feature 1 is `cylinders` and feature 2 is `displacement`.

Perform a test-train split on this reduced column data and call the split data `X_train_reduced`, `X_test_reduced`, `y_test_reduced`, `y_train_reduced`. Use an 80% split.

In [35]:
# Your code here:
auto.columns

Index(['mpg', 'cylinders', 'displacement', 'horse_power', 'weight',
       'acceleration', 'model_year', 'car_name'],
      dtype='object')

In [36]:
X_reduced = auto[['cylinders', 'displacement','acceleration']]
X_train_reduced, X_test_reduced, y_train_reduced, y_test_reduced = train_test_split(X_reduced, y, test_size=0.20)

Generate a new model called `auto_model_reduced` and fit this model. Then proceed to compute the r squared score for the model. Did this cause an improvement in the r squared score?

In [37]:
# Your code here: 
auto_model_reduced = linear_model.LinearRegression()
moto = auto_model_reduced.fit(X_train_reduced,y_train_reduced)
print(moto.coef_)
print(moto.intercept_)

[[-0.67696683 -0.05123187 -0.03499921]]
[37.80326043]


In [38]:
y_moto = moto.predict(X_train_reduced)
r2_score(y_train_reduced, y_moto)

0.6592377512000842

In [39]:
y_test_moto=moto.predict(X_test_reduced)
r2_score(y_test_reduced, y_test_moto)

0.592703528562963

#### Conclusion

You may obtain the impression from this lab that without knowing statistical methods in depth, it is difficult to make major progress in machine learning. That is correct. If you are motivated to become a data scientist, statistics is the subject you must be proficient in and there is no shortcut. 

Completing these labs is not likely to make you a data scientist. But you will have a good sense about what are there in machine learning and what are good for you. In your future career, you can choose one of the three tracks:

* Data scientists who need to be proficient in statistical methods.

* Data engineers who need to be good at programming.

* Data integration specialists who are business or content experts but also understand data and programming. This cross-disciplinary track brings together data, technology, and business and will be in high demands in the next decade.